### ISSUE TITLE
lint checking is slow for considerably big projects

### ISSUE THREAD

**Issue by gokatz** [ember-template-lint/ember-template-lint#534]
`ember-template-lint` (`v1.0.0-beta.1`) is running for a long time compared to previous versions (before using `globby`). After [introducing `globby`](https://github.com/ember-template-lint/ember-template-lint/commit/668a72e49c37b4aa9d15c1789701875bda84e5ab#diff-b458e216195dd6fbfb051f0e3ee47e9bL7), I can sense the delay very well.

I tried with earlier version (0.7.6), there was no lag. However, I can't reproduce using a fresh repo :(

Deps: 
```
"ember-cli": "3.4.3"
"ember-source": "3.4.1"
```

**Comment by rwjblue** [comment]
Interesting! Loads of questions follow 😆:

* Can you measure the time before the global change and after for a rough idea of the delay here (I often use `time ember-template-lint whatever/glob/here` to measure a commands run time)? 
* Can you share how big “big” is? 
* Maybe cloc output or something? 
* What command are you using to run?
* Is it slow when running against a single file, or only when running against all files in the project?

**Comment by gokatz** [comment]
```
1. Can you measure the time before the global change and after for a rough idea of 
the delay here (I often use time ember-template-lint whatever/glob/here to measure 
a commands run time)?
```
**1.0.0-beta.1:**
Full app: around 16s
Single File: arund 8s

**0.7.6:**
Full app: around 1.5s
Single File: arund 0.6s

```
2. Can you share how big “big” is?
3. Maybe cloc output or something?
```
THB, I don't know how to measure 😀
Main app has around 30 routes and 6 engines. 

**Cloc output:**
app/
<img width="598" alt="screen shot 2018-10-04 at 10 43 54 pm" src="https://user-images.githubusercontent.com/15158776/46491014-2849b900-c827-11e8-9131-283a54166f1b.png">

lib/
<img width="570" alt="screen shot 2018-10-04 at 10 43 59 pm" src="https://user-images.githubusercontent.com/15158776/46491021-2da70380-c827-11e8-8024-29b478008323.png">


```
4. What command are you using to run?
```
`yarn ember-template-lint <folderpath>`  and `yarn ember-template-lint <filename>`

```
5. Is it slow when running against a single file, or only when running against all files in the project?
```
both

@rwjblue Hope this infos help 😉Please let me know if you need any other metrics!

**Comment by rwjblue** [comment]
Thank you very much!

**Comment by bendemboski** [comment]
I ran into some very poor performance a little while back and did some investigation (but didn't put those results in an issue or anything 😭 ).

For me the problem was that my local tmp directory had tons and tons of files (which won't happen with new ember-cli). But I think I can reconstruct what I found.

1. `ember-template-lint` [invokes](https://github.com/ember-template-lint/ember-template-lint/blob/master/bin/ember-template-lint.js#L64) `globby.sync` with `gitignore: true`
2. `globby.sync` [invokes](https://github.com/sindresorhus/globby/blob/master/index.js#L102) `gitignore.sync` with `ignore: opts.ignore` (which is empty because nothing was passed in step 1)
3. `globby.sync` [invokes](https://github.com/sindresorhus/globby/blob/master/gitignore.js#L91) `fastglob` to search the entire file tree for `.gitignore` files (adding some default ignored directories to the search, but of course not `tmp`).

This last step is very un-optimized for the case of a big `tmp` folder because it just searches for `.gitignore` files everywhere and then parses them all, rather than running a recursive algorithm that parses each `.gitignore` file and skips recursing into directories that it specifies to ignore. Probably the vast majority of problematic cases would be optimized by just checking the top-level `.gitignore` and using that to ignore folders (rather than implementing a fully recursive algorithm) because I think most projects just use a single top-level `.gitignore`.

So I could see a few solutions that I think would work:

1. In step (1), pass a short hard-coded list of ignored globs (maybe just `/tmp` and `/dist`)
2. In step (1), do our own parsing of the top-level `.gitignore` and use it to pass a list of ignore globs to `globby.sync`.
3. PR a change into `globby` to implement the optimization of parsing the top-level `.gitignore` to filter the full search for `.gitignore` files.

**Comment by bardzusny** [comment]
Is there any update on this issue? It's currently a definite blocker from upgrading large projects to `1.0.0` (beta 6, as of this writing).

Here's my report from upgrade attempt:

I'm using following NPM script for linting all templates in the project:
```js
"lint:hbs": "ember-template-lint app/**"
```

I'm perfectly fine with this, and never felt the need for `.gitignore` patterns to be taken into account. This should apply to almost all Ember projects. With needs of the rest easily solved by passing more-customized glob string as CLI executable param.

Here's analogous issue I've found in ESLint arguing against parsing of `.gitignore`: https://github.com/eslint/eslint/issues/5848 . 

Digressing from this, some numbers from real-life project:
* All application templates:
  * Before `globby`: `npm run lint:hbs  3,19s user 0,14s system 132% cpu 2,519 total`
  * With `globby`: `npm run lint:hbs  77,29s user 28,07s system 112% cpu 1:33,86 total`
* Single template:
  *  Before `globby`: `ember-template-lint app/pods/application/template.hbs  0,14s user 0,02s system 108% cpu 0,154 total`
  * With `globby`: `ember-template-lint app/pods/application/template.hbs  2,68s user 1,43s system 112% cpu 3,642 total`

That's with "*fresh*" `tmp` folder (`rm -r tmp && ember server`). Before I did this, linting one file with `globby` took _over 20s_. Linting whole project took forever (I gave up after waiting around 5-10 minutes). That's on i9 machine. (Again: _20s for single file_ :slightly_smiling_face: .)

Bottom line: can we revert https://github.com/ember-template-lint/ember-template-lint/pull/459 until this problem is sorted out upstream? I take ignoring `.gitignore` over this sort of performance degradation, any day.

**Comment by astronomersiva** [comment]
I believe #607 fixed this as well. Not seeing the slowness after upgrading to v1.1.0 that was released just a couple of days ago.

**Comment by rwjblue** [comment]
Awesome, thank you for confirming!
